### PR TITLE
EthCustodian: make AdminControlled and pausable

### PR DESCRIPTION
**Pull request by sept-en**

### CODE REVIEW

File: eth-custodian/package.json [hunk 1]
```diff
@@ -13,7 +13,7 @@
         "hardhat": "^2.0.8",
         "hardhat-gas-reporter": "^1.0.4",
         "near-api-js": "^0.40.0",
-        "rainbow-bridge": "https://github.com/near/rainbow-bridge",
+        "rainbow-bridge": "https://github.com/near/rainbow-bridge#ede62fcb3614e28e8c5406300381319fec846100",
```
Comment (mfornet) [inline eth-custodian/package.json:16]: Why you depend on a particular version?

Comment (sept-en) [reply eth-custodian/package.json:16]: Rainbow-token-connector [depends](https://github.com/aurora-is-near/rainbow-token-connector/blob/a45e23aed6d3566ea8f89da7d9527da931699fe6/erc20-connector/package.json#L9) on a particular version as well. Moreover, hardhat might fail with a very strange error telling "Missing task argument file" which I figured out can appear when some included contracts are using the newer version of Solidity contract.

File: eth-custodian/contracts/EthCustodian.sol [hunk 2]
```diff
@@ -55,13 +66,16 @@ contract EthCustodian is ProofKeeper {
     function depositToNear(string memory nearRecipientAccountId, uint256 fee)
         external
         payable
+        pausable (PAUSED_DEPOSIT_TO_NEAR)
     {
         require(fee < msg.value, "The fee cannot be bigger than the transferred amount.");
         emit Deposited(msg.sender, nearRecipientAccountId, msg.value, fee);
     }
```
[the remark this answers is not in the document]
Comment (sept-en) [reply eth-custodian/contracts/EthCustodian.sol:74]: This is the near account address (string).

File: eth-custodian/contracts/EthCustodian.sol [hunk 3]
```diff
@@ -1,11 +1,12 @@
 pragma solidity ^0.6.12;
 
-import "rainbow-bridge/contracts/eth/nearprover/contracts/ProofDecoder.sol";
+import "rainbow-bridge/contracts/eth/nearbridge/contracts/AdminControlled.sol";
 import "rainbow-bridge/contracts/eth/nearbridge/contracts/Borsh.sol";
+import "rainbow-bridge/contracts/eth/nearprover/contracts/ProofDecoder.sol";
 
 import { INearProver, ProofKeeper } from "./ProofKeeper.sol";
```
[the remark this answers is not in the document]
Comment (sept-en) [reply eth-custodian/contracts/EthCustodian.sol:7]: The idea here is to use the same `INearProver` that is used in `ProofKeeper`. So if `ProofKeeper` changes the reference contract for the prover after some code update, we won't need to update this here accordingly.

File: eth-custodian/contracts/EthCustodian.sol [hunk 4]
```diff
@@ -40,6 +50,7 @@ contract EthCustodian is ProofKeeper {
     function depositToEVM(string memory ethRecipientOnNear, uint256 fee)
         external
         payable
+        pausable (PAUSED_DEPOSIT_TO_EVM)
     {
         require(fee < msg.value, "The fee cannot be bigger than the transferred amount.");
```
[the remark this answers is not in the document]
Comment (sept-en) [reply eth-custodian/contracts/EthCustodian.sol:55]: As we discussed it some long time ago, should be valid. The `fee` is deducted from the `amount` (`msg.value`) on the Near side so it absolutely shouldn't be bigger. if the `fee` == `msg.value` that means that the fee will be paid but the user will receive 0. That basically makes no sense, also like setting a big fee and transfer a couple/dozens of wei, etc. But this should be valid I think. @mfornet take a look and confirm, please.

@mrLSD I believe we can mint 0 amount having no issues with it on the Near side, right?

File: eth-custodian/contracts/EthCustodian.sol [hunk 5]
```diff
@@ -25,13 +26,22 @@ contract EthCustodian is ProofKeeper {
...
     /// It also links to the prover that it uses to withdraw the tokens.
-    constructor(bytes memory nearEvm, INearProver prover, uint64 minBlockAcceptanceHeight, address _admin)
+    constructor(bytes memory nearEvm,
+                INearProver prover,
+                uint64 minBlockAcceptanceHeight,
+                address _admin,
+                uint pausedFlags)
+        AdminControlled(_admin, pausedFlags)
         ProofKeeper(nearEvm, prover, minBlockAcceptanceHeight)
         public
```
[the remark this answers is not in the document]
Comment (sept-en) [reply eth-custodian/contracts/EthCustodian.sol:43]: According to the Solidity [style-guide](https://docs.soliditylang.org/en/v0.6.0/style-guide.html#function-declaration), visibility modifier comes after the parent constructor call.

File: eth-custodian/test/EthCustodian.js [hunk 6]
```diff
@@ -294,4 +312,435 @@ describe('EthCustodian contract', () => {
...
+            await expect(
+                ethers.provider.sendTransaction(signedTx)
+            )
+                .to
+                .emit(ethCustodian, 'Deposited')
+                .withArgs(walletUser1.address, nearRecipientAccountId, unsignedTx.value, fee);
+
+            const balanceAfter = ethers.BigNumber.from(await ethers.provider.getBalance(ethCustodian.address));
+            const balanceDiff = balanceAfter.sub(balanceBefore);
+            await expect(
```
Comment (abacabadabacaba) [inline eth-custodian/test/EthCustodian.js:345]: Here (and in many other places), the `await` appears to be redundant.

**Review by abacabadabacaba — CHANGES_REQUESTED** [review]

File: eth-custodian/test/EthCustodian.js [hunk 7]
```diff
@@ -29,27 +35,37 @@ describe('EthCustodian contract', () => {
     const nearEvmAccount = Buffer.from('v1.eth-connector.testnet');
 
     beforeEach(async () => {
-        [deployerAccount, ethRecipientOnNear, user2] = await ethers.getSigners();
+        [deployerAccount, ethRecipientOnNear, user2, user3] = await ethers.getSigners();
+
+        const hardhatTestMnemonic = 'test test test test test test test test test test test junk';
+        const derivationPathAdmin = 'm/44\'/60\'/0\'/0/10';
+        walletAdmin = await ethers.Wallet.fromMnemonic(hardhatTestMnemonic, derivationPathAdmin);
```
[the remark this answers is not in the document]
Comment (sept-en) [reply eth-custodian/test/EthCustodian.js:42]: This was needed to explicitly sign the transaction.

File: eth-custodian/test/EthCustodian.js [hunk 8]
```diff
@@ -294,4 +312,435 @@ describe('EthCustodian contract', () => {
...
+            // Let's try to deposit some tokens to Near NEP-141
+            // The balance should be changed and the `Deposited` event emitted
+            const balanceBefore = ethers.BigNumber.from(await ethers.provider.getBalance(ethCustodian.address));
+
+            // Prepare the 1st TX (should succeed)
+            const amountToTransfer = 5000; // wei
+            let unsignedTx = await ethCustodian
+                .connect(walletUser1)
+                .populateTransaction
+                .depositToNear(nearRecipientAccountId, fee);
```
[the remark this answers is not in the document]
Comment (sept-en) [reply eth-custodian/test/EthCustodian.js:330]: Good point! I remember that I needed to implement such a workflow when I implemented the first tests for the contract. I'll check and remove all unnecessary cases if it's redundant already and can be replaced by regular contract call with options provided.

File: eth-custodian/contracts/EthCustodian.sol (hunk 5 above)
[the remark this answers is not in the document]
Comment (0x3bfc) [reply eth-custodian/contracts/EthCustodian.sol:42]: Good to know ... thanks 👍

File: eth-custodian/test/EthCustodian.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (sept-en) [reply eth-custodian/test/EthCustodian.js]: Removed explicitly tx creation and signing for the sake of readability. Thank you for your suggestion!

File: eth-custodian/test/EthCustodian.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (sept-en) [reply eth-custodian/test/EthCustodian.js]: Removed as we got rid of explicit transaction creating&signing.

**Comment by sept-en** [comment]
@abacabadabacaba thank you for the review! your comments were addressed, please check.

**Pull request merged**